clang/test/Sema/inline-asm-validate-spirv-amdgcn.cl

```diff
@@ -0,0 +1,111 @@
+// REQUIRES: amdgpu-registered-target
+// RUN: %clang_cc1 -triple spirv64-amd-amdhsa -fsyntax-only -verify %s
+
+#pragma OPENCL EXTENSION cl_khr_fp64 : enable
+
+kernel void test () {
+
+  int sgpr = 0, vgpr = 0, imm = 0;
+
+  // sgpr constraints
+  __asm__ ("s_mov_b32 %0, %1" : "=s" (sgpr) : "s" (imm) : );
+
+  __asm__ ("s_mov_b32 %0, %1" : "={s1}" (sgpr) : "{exec}" (imm) : );
+  __asm__ ("s_mov_b32 %0, %1" : "={s1}" (sgpr) : "{exe" (imm) : ); // expected-error {{invalid input constraint '{exe' in asm}}
+  __asm__ ("s_mov_b32 %0, %1" : "={s1}" (sgpr) : "{exec" (imm) : ); // expected-error {{invalid input constraint '{exec' in asm}}
+  __asm__ ("s_mov_b32 %0, %1" : "={s1}" (sgpr) : "{exec}a" (imm) : ); // expected-error {{invalid input constraint '{exec}a' in asm}}
```

Comment:
If we really have to tolerate asm, I wonder if we can ban physical register references through SPIRV (other than for maybe the handful of named special register). That is, allow exec/m0/vcc and disallow any numbered registers. The exact register counts move around from target to target and are exposing bonus ABI

Reply:
I think that, whilst appealing, it'd be problematic in a few ways, but mainly it'd relegate this to interesting tech experiment that is not particularly useful to existing users. There's already oodles of AMDGCN ASM that has been authored (a lot of it unguarded to boot); most of it is actually arithmetic, that tickles numbered registers to achieve peak performance. Telling folks that they need to rewrite swathes of their libs / core algorithms in order to get SPIRV is highly likely to ensure they aggressively ignore it. To my mind this is a necessary evil.

Reply:
Or, maybe, we can use this to finally get them to eliminate the asm in the first place

Reply:
We can use non-flavoured SPIRV for that, no? Once the rest of the toolchain bits are in place, that'll work as well, and will offer neither ASM nor builtins. Surely, if there's actual some groundswell of opposition to ASM, folks will organically migrate. I'd say that's preferable to trying to force users to do anything, which seldom works well.